enterprise/cmd/frontend/internal/batches/webhooks/azuredevops.go

```diff
@@ -0,0 +1,186 @@
+package webhooks
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"strconv"
+
+	sglog "github.com/sourcegraph/log"
+
+	fewebhooks "github.com/sourcegraph/sourcegraph/cmd/frontend/webhooks"
+	"github.com/sourcegraph/sourcegraph/enterprise/internal/batches/store"
+	"github.com/sourcegraph/sourcegraph/internal/actor"
+	"github.com/sourcegraph/sourcegraph/internal/database"
+	"github.com/sourcegraph/sourcegraph/internal/extsvc"
+	"github.com/sourcegraph/sourcegraph/internal/extsvc/azuredevops"
+	"github.com/sourcegraph/sourcegraph/internal/gitserver"
+	"github.com/sourcegraph/sourcegraph/internal/types"
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+	"github.com/sourcegraph/sourcegraph/schema"
+)
+
+var azureDevOpsEvents = []string{
+	"git.pullrequest.created",
+	"git.pullrequest.updated",
+	"git.pullrequest.merged",
+}
+
+type AzureDevOpsWebhook struct {
+	*webhook
+}
+
+func NewAzureDevOpsWebhook(store *store.Store, gitserverClient gitserver.Client, logger sglog.Logger) *AzureDevOpsWebhook {
+	return &AzureDevOpsWebhook{
+		webhook: &webhook{store, gitserverClient, logger, extsvc.TypeAzureDevOps},
+	}
+}
+
+func (h *AzureDevOpsWebhook) Register(router *fewebhooks.Router) {
+	router.Register(
+		h.handleEvent,
+		extsvc.KindAzureDevOps,
+		azureDevOpsEvents...,
+	)
+}
+
+func (h *AzureDevOpsWebhook) handleEvent(ctx context.Context, db database.DB, codeHostURN extsvc.CodeHostBaseURL, event any) error {
+	ctx = actor.WithInternalActor(ctx)
+
+	prs, ev, err := h.convertEvent(event)
+	if err != nil {
+		return err
+	}
+
+	for _, pr := range prs {
+		if pr == (PR{}) {
+			h.logger.Warn("Dropping Azure DevOps webhook event", sglog.String("type", fmt.Sprintf("%T", event)))
+			continue
+		}
+
+		eventErr := h.upsertChangesetEvent(ctx, codeHostURN, pr, ev)
+		if eventErr != nil {
+			err = errors.Append(err, eventErr)
+		}
+	}
+	return err
+}
+
+func (h *AzureDevOpsWebhook) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	e, extSvc, hErr := h.parseEvent(r)
+	if hErr != nil {
+		respond(w, hErr.code, hErr)
+		return
+	}
+
+	fewebhooks.SetExternalServiceID(r.Context(), extSvc.ID)
+
+	// 🚨 SECURITY: now that the shared secret has been validated, we can use an
+	// internal actor on the context.
+	ctx := actor.WithInternalActor(r.Context())
+
+	c, err := extSvc.Configuration(ctx)
+	if err != nil {
+		h.logger.Error("Could not decode external service config", sglog.Error(err))
+		http.Error(w, "Invalid external service config", http.StatusInternalServerError)
+		return
+	}
+
+	config, ok := c.(*schema.AzureDevOpsConnection)
+	if !ok {
+		h.logger.Error("Could not decode external service config", sglog.Error(err))
+		http.Error(w, "Invalid external service config", http.StatusInternalServerError)
+		return
+	}
+
+	codeHostURN, err := extsvc.NewCodeHostBaseURL(config.Url)
+	if err != nil {
+		respond(w, http.StatusInternalServerError, errors.Wrap(err, "parsing code host base url"))
```

Comment:
Don't you need a `return` here?
```suggestion
		respond(w, http.StatusInternalServerError, errors.Wrap(err, "parsing code host base url"))
		return
```